### ISSUE TITLE
Support remote include in blockdiagram

### ISSUE THREAD

**Issue by dacofr** [asciidoctor/asciidoctor-intellij-plugin#138]
It will be usefull if the plugin could support the "include::" tag to download external resource

```
[source,json]
----
include::http://infra-config-preprod:9292/environnements/recette[]
----

[graphviz,toto,png]
----
include::http://localhost:9292/graphs/esb_reindex[]
----
```

Exemple : 
![image](https://cloud.githubusercontent.com/assets/1257842/18403192/abcc7c96-76e3-11e6-92d6-8a2a03b06964.png)

AsciiDoc Plugin Version: 0.15.3

IDE Name and Version: IntelliJ IDEA 2016.2.2 EAP
Build #IU-162.1628.17, built on August 8, 2016
JRE: 1.8.0_92-b14 amd64

**Comment by mojavelinux** [comment]
I agree. The `allow-uri-read` would need to be specified on a settings page as it must be passed to the API directly. That attribute is not honored when set in the document itself (given that it's a security-related attribute).

**Comment by ahus1** [comment]
As this is security related - would you expect this to be a IDE or a project setting?

When it is only about editing and previewing, I could provide a snipped that would conditionally load a static image from disk when edited, but loads a remote image when rendered - with different security settings - on the command line. 

It would be helpful as well to know more about the scenario where you plan to use this feature. 

At the moment I'm a bit skeptical about this: if you would open any open source project with an asciidoc document, you'll trigger the included links when previewing them.

**Comment by mojavelinux** [comment]
> As this is security related - would you expect this to be a IDE or a project setting?

Exactly my thoughts. Reaching out to the internet is a security concern, so we want to make sure we can consent from the user. If that consent is given, it can be applied globally (either at the project level or the IDE level).

>  if you would open any open source project with an asciidoc document, you'll trigger the included links when previewing them.

That's exactly why we need to get consent. But once that consent is given, then we should allow it to happen.

**Comment by ahus1** [comment]
@dacofr  - could you please describe more about your scenario? It would be helpful to know more how you plan to use this feature.

**Comment by ahus1** [comment]
@dacofr, I haven't heard from you for two weeks. I hesitate to relax security here without considering other options. For considering other options, I need to know more about what you're trying to achieve. Please feel free to contact me by email if you don't want to discuss this on Github. 

For now I'm closing this issue. It'll be reopened once more information is available.

**Comment by mojavelinux** [comment]
@ahus1 I'd say this is a pretty common request. Users might want to reference diagrams that are stored in other repositories, which is most convenient by URL. It's not a problem to include source from a URL as long as the user gives consent. Is it possible to put this setting in a configuration screen somewhere in IntelliJ?

**Comment by dacofr** [comment]
Hi @ahus1 ,

sorry for my late answer :(

For my use case, i need that the plugin download the external content and then render the content inside my document.  For me it could be an IDE settings.

Thanks for you interest on my issue.

**Comment by ahus1** [comment]
@dacofr - thank you for your reply. I'm reopening this issue. 

I plan to add an option on a per-project basis, but there is still a lot for me to learn how to do this as I don't have experience with IntelliJ UI dialogues. It will take some time.

But I'd like to make sure that this will give you the experience you need: I wonder if you're using the plugin to read documentation (so that is a nice asciidoc-"browser"), or if you want a nice preview when writing the documentation about your production environment.

When writing a document, the rendering of the preview will trigger a fetch of the included resources on every key stroke. The preview might even hang when your remote system is responding slowly. To have a nice (and fast) preview during writing, you could do conditional rendering: when editing you would show a static local (and fast) placeholder, when rendering the document i.e. on a continuous integration server it will fetch the real URLs.

I have some conditional rendering for my Asciidoc documents on Github: When people look at them in the Github Repo, I want them to go to the renedered site. Therefore I add the following text (code?) that will show a link only when rendered in the Github repo browser:

```
ifdef::env-github[]
To read this tutorial fully rendered, please visit: http://....
endif::[]
```

When you are using your IDE for browsing documentation, the preview might still hang, but the resources will only be fetched once - until you start editing your document. And it might to be an option to switch the include with placeholder.

If you would reference SVGs or PNGs by URL, they will not slow down the rendering of the preview. The preview will show the text, and once the images are retrieved they will be shown as well. The preview will even cache images.

Please let me know if you are a "reader" or a "writer" and if a conditional placeholder could work for you.

PS: In real life I might be tempted to the ["five why's"](https://en.wikipedia.org/wiki/5_Whys) - but this would be very strange thing in a Github issue.

**Comment by mojavelinux** [comment]
> When writing a document, the rendering of the preview will trigger a fetch of the included resources on every key stroke.

Asciidoctor has a built-in cache for remote resources. You just need to enable the `cache-uri` attribute on the document. (AsciidoctorJ includes and autoloads the required gem).

Keep in mind that the cache does not timeout, so you'd have to remove the temporary folder manually to force a refresh. (We don't implement invalidation in core).

**Comment by dacofr** [comment]
Hi @ahus1 ,

in fact i'm the two :) But at the moment i'm more a writer.
Like @mojavelinux say you can use the built-cache provided by Asciidoctor.
Adding a "flush cache" button will be great to remove temporary files.

Thanks

**Comment by mojavelinux** [comment]
> Adding a "flush cache" button will be great to remove temporary files.

I don't think we should go there yet because technically that should be something that happens automatically. It's best not to be applying duck tape yet.

**Comment by ahus1** [comment]
As of version 0.21.1 of the plugin (Sep 29, 2018) a user can define attributes like allow-uri-read in the plugin's setting.

I tried this with the most recent version of the plugin (0.28.7), but my example failed to run in the preview with a message "Failed to load AsciiDoc document - no implicit conversion of URI::HTTP into String" when trying to convert for example

````
----
include::http://asciidoc.org/index.txt[]
----
````

although this was working on the command line. Suggestions on how to change this are welcome.

**Comment by ahus1** [comment]
I found the reason for this, it will be fixed as part of #348, I close this issue as a duplicate.

**Comment by ahus1** [comment]
A new pre-release 0.30.10 of this plugin includes this feature. The preview is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://github.com/asciidoctor/asciidoctor-intellij-plugin/blob/master/CONTRIBUTING.adoc#eap).
